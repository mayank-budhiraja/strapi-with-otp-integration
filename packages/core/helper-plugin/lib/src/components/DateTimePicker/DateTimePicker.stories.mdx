<!--- DateTimePicker.stories.mdx --->

import { useState } from 'react';
import { Meta, Story, Canvas } from '@storybook/addon-docs/blocks';
import DateTimePicker from './index';

<Meta title="components/DateTimePicker" component={DateTimePicker} />

# DateTimePicker

This is the doc of the `DateTimePicker` component

## Base

Description...

<Canvas>
  <Story name="base">
    {() => {
      const [value, setValue] = useState();
      return (
        <DateTimePicker
          onClear={() => setValue(undefined)}
          value={value}
          onChange={e => setValue(e)}
          label="Date time picker"
          hint="This is a super description"
        />
      );
    }}
  </Story>
</Canvas>

## Error

Description...

<Canvas>
  <Story name="error">
    <DateTimePicker
      label="Date time picker"
      hint="This is a super description"
      error="Very very very very very very very long error"
    />
  </Story>
</Canvas>

## Initial data

Description...

<Canvas>
  <Story name="intial data">
    <DateTimePicker
      label="Date time picker"
      hint="This is a super description"
      value="2021-10-13T10dd:00:00.000Z"
    />
  </Story>
</Canvas>
