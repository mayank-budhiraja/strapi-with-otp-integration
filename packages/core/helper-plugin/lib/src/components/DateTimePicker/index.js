import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Stack } from '@strapi/design-system/Stack';
import { DatePicker } from '@strapi/design-system/DatePicker';
import { TimePicker } from '@strapi/design-system/TimePicker';
import { Field, FieldHint, FieldLabel, FieldError } from '@strapi/design-system/Field';
import { Flex } from '@strapi/design-system/Flex';
import { Box } from '@strapi/design-system/Box';
import { parseDate } from './parseDate';

const CustomField = styled(Field)`
  > div {
    > div {
      p[data-strapi-field-error='true'] {
        display: none;
      }
    }
  }
`;

const LabelAction = styled(Box)`
  svg path {
    fill: ${({ theme }) => theme.colors.neutral500};
  }
`;

const DateTimePicker = ({
  ariaLabel,
  clearLabel,
  disabled,
  error,
  hint,
  label,
  labelAction,
  onChange,
  onClear,
  name,
  required,
  size,
  step,
  value,
  ...props
}) => {
  const initialDate = parseDate(value);
  const [dateValue, setDateValue] = useState(initialDate);
  const [timeValue, setTimeValue] = useState(
    initialDate
      ? `${initialDate.getHours()}:${initialDate.getMinutes()}:${initialDate.getSeconds()}`
      : null
  );

  const handleDateChange = e => {
    setDateValue(e);

    if (timeValue) {
      const dateToSet = new Date(e);
      dateToSet.setHours(timeValue.split(':')[0]);
      dateToSet.setMinutes(timeValue.split(':')[1]);

      if (onChange) {
        onChange(dateToSet);
      }
    } else {
      const dateToSet = new Date(e);
      setTimeValue(`${dateToSet.getHours()}:${dateToSet.getMinutes()}:${dateToSet.getSeconds()}`);

      if (onChange) {
        onChange(dateToSet);
      }
    }
  };

  const handleTimeChange = e => {
    setTimeValue(e);

    const dateToSet = new Date(dateValue);
    dateToSet.setHours(e.split(':')[0]);
    dateToSet.setMinutes(e.split(':')[1]);

    if (!dateValue) {
      setDateValue(new Date());
    }

    if (onChange) {
      onChange(dateToSet);
    }
  };

  const handleDateClear = () => {
    setDateValue(undefined);
    setTimeValue(undefined);
    onClear();
  };

  const handleTimeClear = () => {
    setTimeValue(undefined);

    let dateToSet;

    if (dateValue) {
      dateToSet = new Date(dateValue);
      dateToSet.setHours('00');
      dateToSet.setMinutes('00');
    }

    if (onChange) {
      onChange(dateToSet);
    }
  };

  return (
    <CustomField
      name={name}
      role="group"
      aria-labelledby="datetime-label"
      hint={hint}
      error={error}
    >
      <Stack spacing={1}>
        {label && (
          <Flex>
            <FieldLabel required={required} id="datetime-label">
              {label}
            </FieldLabel>
            {labelAction && <LabelAction paddingLeft={1}>{labelAction}</LabelAction>}
          </Flex>
        )}
        <Stack horizontal spacing={2}>
          <DatePicker
            data-testid="datetimepicker-date"
            name={name}
            ariaLabel={label || ariaLabel}
            error={error}
            selectedDate={dateValue}
            selectedDateLabel={formattedDate => `Date picker, current is ${formattedDate}`}
            onChange={handleDateChange}
            size={size}
            onClear={onClear && handleDateClear}
            clearLabel={clearLabel}
            disabled={disabled}
            {...props}
          />
          <TimePicker
            data-testid="datetimepicker-time"
            size={size}
            aria-label={label || ariaLabel}
            error={error}
            value={timeValue}
            onChange={handleTimeChange}
            onClear={onClear && handleTimeClear}
            clearLabel={clearLabel}
            disabled={disabled}
            step={step}
          />
        </Stack>
        <FieldHint />
        <FieldError id="datetimepicker" />
      </Stack>
    </CustomField>
  );
};

DateTimePicker.displayName = 'DateTimePicker';

DateTimePicker.defaultProps = {
  ariaLabel: undefined,
  clearLabel: 'clear',
  disabled: false,
  error: undefined,
  hint: undefined,
  label: undefined,
  labelAction: undefined,
  onClear: undefined,
  required: false,
  size: 'M',
  step: 1,
  value: undefined,
};

DateTimePicker.propTypes = {
  ariaLabel: PropTypes.string,
  clearLabel: PropTypes.string,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  hint: PropTypes.string,
  label: PropTypes.string,
  labelAction: PropTypes.element,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  onClear: PropTypes.func,
  required: PropTypes.bool,
  size: PropTypes.oneOf(['S', 'M']),
  step: PropTypes.number,
  value: PropTypes.instanceOf(Date),
};

export default DateTimePicker;
