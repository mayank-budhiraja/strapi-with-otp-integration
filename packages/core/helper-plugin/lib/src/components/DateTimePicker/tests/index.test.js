import * as React from 'react';
import { render, screen } from '@testing-library/react';
import { ThemeProvider } from '@strapi/design-system/ThemeProvider';
import { lightTheme } from '@strapi/design-system/themes';
import DateTimePicker from '../index';

describe('DateTimePicker', () => {
  it('snapshots the component', () => {
    const { container } = render(
      <ThemeProvider theme={lightTheme}>
        <DateTimePicker
          value={new Date('2021-10-13T10:00:00.000Z')}
          onChange={() => {}}
          name="datetimepicker"
          label="Date time picker"
          hint="This is a super description"
        />
      </ThemeProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should initialize the two inputs', () => {
    render(
      <ThemeProvider theme={lightTheme}>
        <DateTimePicker
          value={new Date('2021-10-13T13:43:00.000Z')}
          step={15}
          onChange={() => {}}
          name="datetimepicker"
          label="Date time picker"
          hint="This is a super description"
        />
      </ThemeProvider>
    );

    expect(screen.getByText('13:45')).toBeInTheDocument();
    const datepicker = screen.getByTestId('datetimepicker-date');

    expect(datepicker.value).toBe('10/13/2021');
  });
});
