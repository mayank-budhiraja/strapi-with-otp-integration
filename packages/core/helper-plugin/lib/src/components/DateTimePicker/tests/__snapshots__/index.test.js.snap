// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimePicker snapshots the component 1`] = `
.c1 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c2 > * {
  margin-top: 0;
  margin-bottom: 0;
}

.c2 > * + * {
  margin-top: 4px;
}

.c7 > * {
  margin-left: 0;
  margin-right: 0;
}

.c7 > * + * {
  margin-left: 8px;
}

.c12 {
  padding-right: 8px;
  padding-left: 12px;
}

.c8 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c10 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c14 {
  border: none;
  border-radius: 4px;
  padding-bottom: 0.65625rem;
  padding-left: 0;
  padding-right: 16px;
  padding-top: 0.65625rem;
  color: #32324d;
  font-weight: 400;
  font-size: 0.875rem;
  display: block;
  width: 100%;
  background: inherit;
}

.c14::-webkit-input-placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c14::-moz-placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c14:-ms-input-placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c14::placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c14[aria-disabled='true'] {
  color: inherit;
}

.c14:focus {
  outline: none;
  box-shadow: none;
}

.c11 {
  border: 1px solid #dcdce4;
  border-radius: 4px;
  background: #ffffff;
  outline: none;
  box-shadow: 0;
  -webkit-transition-property: border-color,box-shadow,fill;
  transition-property: border-color,box-shadow,fill;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
}

.c11:focus-within {
  border: 1px solid #4945ff;
  box-shadow: #4945ff 0px 0px 0px 2px;
}

.c9 > * {
  margin-top: 0;
  margin-bottom: 0;
}

.c9 > * + * {
  margin-top: 4px;
}

.c13 {
  border: none;
  background: transparent;
  border-radius: 4px;
}

.c13 svg path {
  fill: #8e8ea9;
}

.c19 {
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  top: 0;
  width: 100%;
  background: transparent;
  border: none;
}

.c19:focus {
  outline: none;
}

.c19[aria-disabled='true'] {
  cursor: not-allowed;
}

.c22 {
  padding-left: 12px;
}

.c24 {
  padding-right: 16px;
  padding-left: 16px;
}

.c15 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c17 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c20 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c25 {
  color: #32324d;
  display: block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: 0.875rem;
  line-height: 1.43;
}

.c16 > * {
  margin-top: 0;
  margin-bottom: 0;
}

.c18 {
  position: relative;
  border: 1px solid #dcdce4;
  padding-right: 12px;
  border-radius: 4px;
  background: #ffffff;
  overflow: hidden;
  min-height: 2.5rem;
  outline: none;
  box-shadow: 0;
  -webkit-transition-property: border-color,box-shadow,fill;
  transition-property: border-color,box-shadow,fill;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
}

.c18:focus-within {
  border: 1px solid #4945ff;
  box-shadow: #4945ff 0px 0px 0px 2px;
}

.c26 {
  background: transparent;
  border: none;
  position: relative;
  z-index: 1;
}

.c26 svg {
  height: 0.6875rem;
  width: 0.6875rem;
}

.c26 svg path {
  fill: #666687;
}

.c27 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: none;
  border: none;
}

.c27 svg {
  width: 0.375rem;
}

.c21 {
  width: 100%;
}

.c23 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c23 svg {
  height: 1rem;
  width: 1rem;
}

.c23 svg path {
  fill: #8e8ea9;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c4 {
  font-weight: 600;
  color: #32324d;
  font-size: 0.75rem;
  line-height: 1.33;
}

.c28 {
  color: #666687;
  font-size: 0.75rem;
  line-height: 1.33;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 > div > div p[data-strapi-field-error='true'] {
  display: none;
}

<div
  aria-labelledby="datetime-label"
  class="c0"
  role="group"
>
  <div
    class="c1 c2"
    spacing="1"
  >
    <div
      class="c3"
    >
      <label
        class="c4"
        for="field-1"
        id="datetime-label"
      >
        <div
          class="c5"
        >
          Date time picker
        </div>
      </label>
    </div>
    <div
      class="c6 c7"
      spacing="2"
    >
      <div
        class=""
      >
        <div>
          <div>
            <div
              class="c8 c9"
              spacing="1"
            >
              <div
                class="c10 c11"
              >
                <div
                  class="c12"
                >
                  <button
                    aria-disabled="false"
                    aria-label="Date picker, current is 10/13/2021"
                    class="c13"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      fill="none"
                      height="1em"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M3.869 2.99V0h2.9v2.99h10.463V0h2.9v2.99h.629c2.768 0 3.203.498 3.239 2.926V21c0 2.124-.191 3-2.802 3H2.818C.208 24 0 23.363 0 20.785V6.21c.035-2.049.233-3.22 3.001-3.22h.868zM2.32 20.369c0 .811.245.865.776.865h17.905c.53 0 .68-.012.68-.825V8.233c-.015-.627-.219-.737-.631-.737H2.907c-.413 0-.592.09-.587.573v12.3z"
                        fill="#212134"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </button>
                </div>
                <input
                  aria-autocomplete="none"
                  aria-disabled="false"
                  aria-invalid="false"
                  aria-label="Date time picker"
                  class="c14"
                  data-testid="datetimepicker-date"
                  id="datepicker-1"
                  name="datetimepicker"
                  placeholder="1/1/1970"
                  value="10/13/2021"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="c15 c16"
        >
          <div
            class="c17 c18"
          >
            <button
              aria-disabled="false"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-label="Date time picker"
              aria-labelledby="timepicker-1-label timepicker-1-content"
              class="c19"
              data-testid="datetimepicker-time"
              id="timepicker-1"
              type="button"
            />
            <div
              class="c20 c21"
            >
              <div
                class="c17"
              >
                <div
                  aria-hidden="true"
                  class="c22"
                >
                  <div
                    class="c23"
                  >
                    <svg
                      fill="none"
                      height="1em"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M0 12C0 5.37 5.364 0 11.994 0S24 5.37 24 12s-5.376 12-12.006 12S0 18.63 0 12zm2.4 0c0 5.304 4.296 9.6 9.6 9.6 5.304 0 9.6-4.296 9.6-9.6 0-5.304-4.296-9.6-9.6-9.6A9.597 9.597 0 002.4 12zm8.4-6h1.8v6.3l5.4 3.204-.906 1.476L10.8 13.2V6z"
                        fill="#212134"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="c24"
                >
                  <span
                    class="c25"
                    id="timepicker-1-content"
                  >
                    10:00
                  </span>
                </div>
              </div>
              <div
                class="c17"
              >
                <button
                  aria-hidden="true"
                  class="c22 c26 c27"
                  tabindex="-1"
                  type="button"
                >
                  <svg
                    fill="none"
                    height="1em"
                    viewBox="0 0 14 8"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M14 .889a.86.86 0 01-.26.625L7.615 7.736A.834.834 0 017 8a.834.834 0 01-.615-.264L.26 1.514A.861.861 0 010 .889c0-.24.087-.45.26-.625A.834.834 0 01.875 0h12.25c.237 0 .442.088.615.264a.86.86 0 01.26.625z"
                      fill="#32324D"
                      fill-rule="evenodd"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <p
      class="c28"
      id="field-1-hint"
    >
      This is a super description
    </p>
  </div>
</div>
`;
